### PR TITLE
serve the status of image acquiring

### PR DESCRIPTION
**Pull request by enoodle**
When image-inspector fails to acquire an image and is in the serve mode it fails to serve this status and just exists. I consider this a bug.
To fix I added two fields to the metadata status: ImageAcquireSuccess, ImageAcquireError to report that status.

### CODE REVIEW

**Comment by bazulay** [comment]
@simon3z PTAL

**Comment by simon3z** [comment]
@pweil- @ilackarms can you review this?

**Comment by enoodle** [comment]
@ilackarms @pweil-  ping

**Comment by simon3z** [comment]
@enoodle anything that we can test here?

**Comment by enoodle** [comment]
@simon3z This is moving a bulk of code into the `acquireImage` function and then depending on the error either continue or serve the result.
To test this I wanted a good way to fail `acquireImage` and this is one of the reasons that made me create #87.
I can try to move those tests form there to here, but if this is the only problem with this PR then I think we should first go with #87 and then get those tests .

**Comment by simon3z** [comment]
IIUC you want to add the ability to expose a more granular image-inspector status through the rest-api, and you want to do that before a new release.

Now the question is, do you feel more confident that either this PR or #87 are going to solve more issues than creating new ones? If so how can we be sure about that?

Do you prefer to go with #87 with many more changes and partly tested (and you can add more), or do you want to work more on this one with less changes (smaller impact overall) and trying to add some tests?

**Comment by enoodle** [comment]
I feel that the code is getting too much complicated to keep in one package and that #87 simplifies things. most of the code was just moved and not changed so it not such a big change in my opinion.
I think postponing the change will only make it harder later and that fixing problems and getting new releases is not a difficulty worth not making changes for. So I would go with #87 first.

**Comment by enoodle** [comment]
@simon3z I added tests here. Some even helped identify a bug with `decodeDockerResponse` that I fixed.

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -649,3 +606,79 @@ func createOutputDir(dirName string, tempName string) (string, error) {
 	}
 	return dirName, nil
 }
+
+// acquireImage returns error, the docker image that was acquired, the image ID,
+// the container ID that the image was acquired from, and iiapi.FilesFilter for this image.
+func (i *defaultImageInspector) acquireImage(client DockerRuntimeClient) (error, docker.Image, string, string, iiapi.FilesFilter) {
```
Comment (simon3z) [inline pkg/inspector/image-inspector.go]: @enoodle when you return such a large number of items you may want to consider an object approach, e.g. either you make those attributes of the relevant `defaultImageInspector` class (if it makes sense), or you return an object of a new class.

File: pkg/api/types.go [hunk 2]
```diff
@@ -93,6 +93,9 @@ var (
 
 // InspectorMetadata is the metadata type with information about image-inspector's operation
 type InspectorMetadata struct {
+	ImageAcquireSuccess bool   // status of aquiring the image
```
Comment (simon3z) [inline pkg/api/types.go]: @enoodle do you need to make this an attribute or could it be just a method that that relies on whether `ImageAcquireError` is `nil` or not?

File: pkg/api/types.go [hunk 3]
```diff
@@ -93,6 +93,9 @@ var (
...
+	ImageAcquireError   string // error message from aquiring the image
```
Comment (simon3z) [inline pkg/api/types.go]: @enoodle I think this should be an `error` that you stringfy only on when some output requires it

File: pkg/inspector/image-inspector.go [hunk 4]
```diff
@@ -126,131 +136,70 @@ func (i *defaultImageInspector) Inspect() error {
...
+			if err := i.postResults(scanResults); err != nil {
+				log.Printf("Error posting results: %v", err)
+				return nil
+			}
 		}
 	}
 
 	if i.imageServer != nil {
-		return i.imageServer.ServeImage(&i.meta, i.opts.DstPath, scanResults, scanReport, htmlScanReport)
+		if i.meta.ImageAcquireSuccess {
```
Comment (simon3z) [inline pkg/inspector/image-inspector.go]: @enoodle it's not easy to understand why you need this check now. In general I would try to encapsulate any further logic in `ServeImage`.

File: pkg/api/types.go (hunk 3 above)
Comment (enoodle) [reply pkg/api/types.go]: @simon3z This struct is served as a JSON to the user [1], so I can't have `error` here.

[1] https://github.com/openshift/image-inspector/blob/master/pkg/imageserver/webdav.go#L96

File: pkg/api/types.go (hunk 2 above)
Comment (enoodle) [reply pkg/api/types.go]: I can do that, It is mostly effecting on the user side.

File: pkg/inspector/image-inspector.go (hunk 4 above)
Comment (enoodle) [reply pkg/inspector/image-inspector.go]: Will move to the `ServeImage` function.

File: pkg/api/types.go (hunk 3 above)
Comment (enoodle) [reply pkg/api/types.go]: never mind  that, it can json marshal `error`, I will make this change.

Comment (enoodle) [reply pkg/api/types.go]: Sorry, apparently it is not possible to meaningfully do it yet
https://github.com/golang/go/issues/5161

**Comment by enoodle** [comment]
@simon3z PTAL

File: pkg/inspector/image-inspector.go [hunk 5]
```diff
@@ -107,152 +116,108 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
...
-			return err
-		}
-
-		imageMetadata, err := i.createAndExtractImage(client, randomName)
-		if err != nil {
-			return err
-		}
-		i.meta.Image = *imageMetadata
-		scanResults.ImageID = i.meta.Image.ID
+		i.meta.ImageAcquireError = err.Error()
```
Comment (simon3z) [inline pkg/inspector/image-inspector.go]: @enoodle please keep the logic of this (already too long) method simple and return on error.

If you need to serve even on error please use a wrapper method,

File: pkg/inspector/image-inspector.go [hunk 6]
```diff
@@ -649,3 +622,81 @@ func createOutputDir(dirName string, tempName string) (string, error) {
 	}
 	return dirName, nil
 }
+
+// acquireImage returns error and iiapi.FilesFilter for this image.
+func (i *defaultImageInspector) acquireImage(client DockerRuntimeClient) (error, iiapi.FilesFilter) {
+	var err error
+	if len(i.opts.Container) == 0 {
```
Comment (simon3z) [inline pkg/inspector/image-inspector.go]: @enoodle this method calls for a refactoring, anything that looks like

```Golang
func ... {
    if (...) {
        ...
    } else {
        ...
    }
}
```

should be better designed.

Anyway it's not for this PR.

Comment (simon3z) [reply pkg/inspector/image-inspector.go]: @enoodle to be clear you can just factor out those large bodies to other functions, etc. what I am saying is that a whole long function with two in-line code paths should be refactored.

**Pull request merged**